shared/sbl.js

```diff
@@ -0,0 +1,113 @@
+'use strict'
+
+type TypeFilter = (domain: string, selector: string, data: Array<*>) => ?boolean
+type TypeDomain = {
+  selectors: {[string]: Function},
+  domainFilters: Array<TypeFilter>,
+  selectorFilters: {[string]: TypeFilter},
+  data: Object
+}
+
+var domains: {[string]: TypeDomain} = {}
+var globalFilters: Array<TypeFilter> = []
+var pendingCalls = {} // allows sbl to be called regardless of module load order
+
+function sbl (path: string, ...data?: Array<*>) {
+  var [, name, selector] = /^([^/]+)(.+)/.exec(path)
+  var domain = domains[name]
+  if (!domain) {
+    // a call might happen to a selector for domains that haven't been registered
+    // yet, due to module load order. So we save them and run them later.
+    if (!pendingCalls[name]) pendingCalls[name] = []
+    pendingCalls[name].push([path, data])
+    return
+  }
+  // Filters can perform additional functions, and by returning `false` they
+  // can prevent the execution of a selector.
+  for (let gf of globalFilters) {
+    if (gf(name, selector, data) === false) return
+  }
+  for (let df of domain.domainFilters) {
+    if (df(name, selector, data) === false) return
+  }
+  var sf = domain.selectorFilters[selector]
+  if (sf && sf(name, selector, data) === false) return
+  return domain.selectors[selector].call(domain, ...data)
+}
+
+sbl.registerDomain = function (domain: string, selectors: {[string]: Function}) {
+  if (domains[domain]) throw new Error(`${domain} exists`)
+  domains[domain] = {selectors, data: {}, domainFilters: [], selectorFilters: {}}
+  // run any pendingCalls for this domain
+  if (pendingCalls[domain]) {
+    for (var [path, data] of pendingCalls[domain]) {
+      sbl(path, ...data)
+    }
+    delete pendingCalls[domain] // see sbl.throwIfPending
+  }
+}
+
+sbl.addDomainFilter = function (domain: string, filter: TypeFilter) {
+  domains[domain].domainFilters.push(filter)
+}
+
+sbl.setSelectorFilter = function (domain: string, selector: string, filter: TypeFilter) {
+  if (domains[domain].selectorFilters[selector]) {
+    throw new Error(`${domain}/${selector} filter exists`)
+  }
+  domains[domain].selectorFilters[selector] = filter
+}
+
+sbl.addGlobalFilter = function (filter: TypeFilter) {
+  globalFilters.push(filter)
+}
+
+// call at end of sbl setup, right before main run loop of app
+sbl.throwIfPending = function () {
```

Comment:
I do not see how this helps differentiates a valid call from a typo. The way its being used now,  it only checks at application start calls could be invalid at any point in the application's life-cycle(?). Further, if we make a call to an api we expect to be synchronous (unless that call expects no return data) then saving a pending call to be run later also suggests that the application has all async calls. Actually even in the case of async calls when a domain is not found we save a pendingCall and  we just return the function rather than a promise.This suggests API's are write only or event emitters that can manage their own data passage? Its a call with no response but potential action. I think it would be better if you return an error at the time of calling a missing api because that not only identifies whether or not its a typo but also tells where the typo was created without having to encode a complex error text from an object. Also, it  gives notice at the point of an unserviceable request. That way the caller could react and manage its own behavior. If there is a potential for a caller to make a valid call before the state of the application is loaded then there should be an opportunity to check its state and wait for it.